beacon/primitives/state.go

```diff
@@ -260,6 +280,62 @@ func (s *State) UpdateValidatorStatus(index uint32, status uint64, c *config.Con
 	return nil
 }
 
+// UpdateValidatorRegistry updates the registry and updates validator pending activation or exit.
+func (s *State) UpdateValidatorRegistry(c *config.Config) {
+	activeValidatorIndices := GetActiveValidatorIndices(s.ValidatorRegistry)
+	totalBalance := s.GetTotalBalance(activeValidatorIndices, c)
+	maxBalanceChurn := c.MaxDeposit * config.UnitInCoin
+	if maxBalanceChurn < (totalBalance / (2 * c.MaxBalanceChurnQuotient)) {
+		maxBalanceChurn = totalBalance / (2 * c.MaxBalanceChurnQuotient)
+	}
+
+	balanceChurn := uint64(0)
+	for idx, validator := range s.ValidatorRegistry {
+		index := uint32(idx)
+		if validator.Status == PendingActivation && s.ValidatorBalances[index] >= c.MaxDeposit*config.UnitInCoin {
+			balanceChurn += s.GetEffectiveBalance(index, c)
+			if balanceChurn > maxBalanceChurn {
+				break
+			}
+
+			s.UpdateValidatorStatus(index, Active, c)
+		}
+	}
+
+	balanceChurn = 0
+	for idx, validator := range s.ValidatorRegistry {
+		index := uint32(idx)
+		if validator.Status == ActivePendingExit {
+			balanceChurn += s.GetEffectiveBalance(index, c)
+			if balanceChurn > maxBalanceChurn {
+				break
+			}
+
+			s.UpdateValidatorStatus(index, ExitedWithoutPenalty, c)
+		}
+	}
+
+	periodIndex := s.Slot / c.CollectivePenaltyCalculationPeriod
+	totalPenalties := s.LatestPenalizedExitBalances[periodIndex]
+	if periodIndex >= 1 {
+		totalPenalties += s.LatestPenalizedExitBalances[periodIndex-1]
+	}
+	if periodIndex >= 2 {
```

Comment:
No 'else'? So if `periodIndex >= 2`, `totalPenalties` adds twice by indexes of `periodIndex - 1` and `periodIndex - 2`? Is it intended?

Reply:
yes - https://github.com/ethereum/eth2.0-specs/blob/8f29bcdf081fc8b4c731d37e6bb1ade34078c0a3/specs/core/0_beacon-chain.md#validator-registry